src/Npgsql/Internal/EncryptionHandler.cs

```diff
@@ -0,0 +1,42 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Npgsql.Properties;
+using Npgsql.Util;
+
+namespace Npgsql.Internal;
+
+interface IEncryptionHandler
+{
+    bool SupportEncryption { get; }
+
+    Task NegotiateEncryption(NpgsqlConnector connector, SslMode sslMode, NpgsqlTimeout timeout, bool async, bool isFirstAttempt);
+
+    Task AuthenticateSASL(NpgsqlConnector connector, List<string> mechanisms, string username, bool async,
+        CancellationToken cancellationToken);
+}
+
+sealed class EncryptionHandler : IEncryptionHandler
+{
+    public bool SupportEncryption => true;
+
+    public Task NegotiateEncryption(NpgsqlConnector connector, SslMode sslMode, NpgsqlTimeout timeout, bool async, bool isFirstAttempt)
+        => connector.NegotiateEncryption(sslMode, timeout, async, isFirstAttempt);
+
+    public Task AuthenticateSASL(NpgsqlConnector connector, List<string> mechanisms, string username, bool async,
+        CancellationToken cancellationToken)
+        => connector.AuthenticateSASL(mechanisms, username, async, cancellationToken);
+}
+
+sealed class EmptyEncryptionHandler : IEncryptionHandler
```

Comment:
nit: can lift the implementations up to IEncryptionHandler (DIM or just a base class)

Reply:
Not to the interface, but to the base class, sure (though I liked the interface a bit more).